### PR TITLE
argocd-util should allow editing project policies in bulk 

### LINKED ISSUE #2614: argocd-util should allow editing project policies in bulk
# Summary

The utility cli is required to backfill existing projects with `actions/*` action.

# Motivation

 This is required to resolve backward-incompatible change introduced by https://github.com/argoproj/argo-cd/issues/2002.  Although this is a one-time operation it might be useful for community users who upgraded to v1.3 and might be useful in future to back-fill project roles.

### PR DESCRIPTION
**Pull request by alexmt**
Closes #2614 

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [x] Optional. My organization is added to the README.
* [x] I've signed the CLA and my build is green ([troubleshooting builds](https://argoproj.github.io/argo-cd/developer-guide/ci/)). 

CLI example:

```
# add permission to execute any action to `deployer` role in all projects
argocd-util projects update-role-policy '*' set 'action/*' --resource applications --scope '*' --permission allow --dry-run=false --role  '*deployer*'
```

```
# remove permission to override resources from `deployer` role in all projects
argocd-util projects update-role-policy '*' remove override --role '*deployer*'
```

### CODE REVIEW

File: cmd/argocd-util/main.go [hunk 1]
```diff
@@ -220,6 +221,7 @@ func NewImportCommand() *cobra.Command {
 				os.Exit(1)
 			}
 			config, err := clientConfig.ClientConfig()
+			errors.CheckError(err)
```
Comment (alexec) [inline cmd/argocd-util/main.go]: day-0 bug?

Comment (alexmt) [reply cmd/argocd-util/main.go]: yeah, I think check was missing from the beginning.

File: cmd/argocd-util/projects.go [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+
+  # Remove a policy that which manages running (action/*) from all roles which name matches *deployer* in all projects
+  argocd-util projects update-role-policy '*' remove override --role '*deployer*'
+`,
+		Run: func(c *cobra.Command, args []string) {
+			if len(args) != 3 {
+				c.HelpFunc()(c, args)
+				os.Exit(1)
+			}
+			projPattern := args[0]
```
Comment (alexec) [inline cmd/argocd-util/projects.go]: rename `projectSelector`

File: cmd/argocd-util/projects.go [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+	var (
+		clientConfig clientcmd.ClientConfig
+		resource     string
+		scope        string
+		rolePattern  string
+		permission   string
+		dryRun       bool
+	)
+	var command = &cobra.Command{
+		Use:   "update-role-policy PROJECT_SELECTOR ROLENAME_PATTERN MODIFICATION ACTION",
```
Comment (alexec) [inline cmd/argocd-util/projects.go]: this line is wrong, I think it should be:

```
update-role-policy PROJECT_GLOB MODIFICATION ACTION
```

File: cmd/argocd-util/projects.go (hunk 2 above)
Comment (alexec) [inline cmd/argocd-util/projects.go]: 1. Check `len(args) == `
2. Rename to `projectGlob`

File: cmd/argocd-util/projects.go [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+			appclients := appclientset.NewForConfigOrDie(config)
+			projects, err := appclients.ArgoprojV1alpha1().AppProjects(namespace).List(v1.ListOptions{})
+			errors.CheckError(err)
+
+			for _, proj := range projects.Items {
+				if !globMatch(projPattern, proj.Name) {
+					continue
+				}
+				origProj := proj.DeepCopy()
+				updated := false
```
Comment (alexec) [inline cmd/argocd-util/projects.go]: I don't think you should check to see if it is changed, I think you should just update.

File: cmd/argocd-util/projects.go [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+				for i, role := range proj.Spec.Roles {
```
Comment (alexec) [inline cmd/argocd-util/projects.go]: there is a lot of logic in this loop - I think you should extract and write tests for it

**Review by alexec — DISMISSED** [review]

File: cmd/argocd-util/projects.go (hunk 4 above)
Comment (alexmt) [reply cmd/argocd-util/projects.go]: Even if project has no changes update takes some time. If we skip updating the project with no changes then backfill operation will be much slower. I think this check makes sense.

File: cmd/argocd-util/projects.go (hunk 5 above)
Comment (alexmt) [reply cmd/argocd-util/projects.go]: done

File: cmd/argocd-util/projects.go (hunk 3 above)
Comment (alexmt) [reply cmd/argocd-util/projects.go]: nice catch. fixed

**Comment by alexmt** [comment]
@alexec can you take a look again please. your comments were addressed

File: cmd/argocd-util/projects.go [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+		}
+		origProj := proj.DeepCopy()
+		updated := false
+		for i, role := range proj.Spec.Roles {
+			if !globMatch(rolePattern, role.Name) {
+				continue
+			}
+			actionPolicyIndex := -1
+			for i := range role.Policies {
+				parts := split(role.Policies[i], ",")
```
Comment (alexec) [inline cmd/argocd-util/projects.go]: untested?

File: cmd/argocd-util/projects.go [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+			}
+			policyPermission := modification(proj.Name, action)
+			if actionPolicyIndex == -1 && policyPermission != "" {
+				updated = true
+				role.Policies = append(role.Policies, formatPolicy(proj.Name, role.Name, policyPermission))
+			} else if actionPolicyIndex > -1 && policyPermission == "" {
+				updated = true
+				role.Policies = append(role.Policies[:actionPolicyIndex], role.Policies[actionPolicyIndex+1:]...)
+			} else if actionPolicyIndex > -1 && policyPermission != "" {
+				updated = true
```
Comment (alexec) [inline cmd/argocd-util/projects.go]: untested?

**Review by alexec — COMMENTED** [review]
Some minor comments - but I think we should sit down and discus this, as I don't fully understand

**Review by alexec — CHANGES_REQUESTED** [review]
Ops. I believe I request changes for this.

**Review by alexec — APPROVED** [review]
Conditional approval: please create an issue to complete the tests for this.

**Pull request merged**